File: clojure/robot-simulator/_iterations/2/src/robot_simulator.clj
(ns robot-simulator)

(def ^:private bearings
  {:north {:x 0  :y 1  :left :west  :right :east}
   :east  {:x 1  :y 0  :left :north :right :south}
   :south {:x 0  :y -1 :left :east  :right :west}
   :west  {:x -1 :y 0  :left :south :right :north}})

(defn robot [coordinates bearing]
  {:coordinates coordinates :bearing bearing})

(defn turn-right [bearing]
  (:right (bearing bearings)))

(defn turn-left [bearing]
  (:left (bearing bearings)))

(defn- pt+ [{ax :x ay :y} {bx :x by :y}]
  {:x (+ ax bx) :y (+ ay by)})

(defn- apply-instruction [{:keys [coordinates bearing] :as robot} instruction]
  (case instruction
    \R (update robot :bearing turn-right)
    \L (update robot :bearing turn-left)
    \A (assoc robot :coordinates (pt+ coordinates (bearing bearings)))
    (assert false (str "Invalid instruction: " instruction))))

(defn simulate [instructions robot]
  (reduce apply-instruction robot instructions))

File: ruby/robot-simulator/_iterations/2/robot_simulator.rb
class Pt
  attr_reader :x, :y

  def initialize(x, y)
    @x = x
    @y = y
  end

  def +(pt)
    Pt.new(x + pt.x, y + pt.y)
  end

  def to_a
    [x, y]
  end
end

class Robot
  BEARINGS = {
    :east => Pt.new(1, 0),
    :west => Pt.new(-1, 0),
    :north => Pt.new(0, 1),
    :south => Pt.new(0, -1)
  }
  CLOCKWISE_BEARINGS = %i[north east south west]

  attr_reader :pt, :bearing

  def initialize
    @pt = Pt.new(0, 0)
    @bearing = :north
  end

  def orient(direction)
    raise ArgumentError, "Invalid bearing: #{direction}" unless BEARINGS.has_key? direction
    @bearing = direction
  end

  def turn_right
    @bearing = CLOCKWISE_BEARINGS[(CLOCKWISE_BEARINGS.index(bearing) + 1) % CLOCKWISE_BEARINGS.size]
  end

  def turn_left
    @bearing = CLOCKWISE_BEARINGS[(CLOCKWISE_BEARINGS.index(bearing) + CLOCKWISE_BEARINGS.size - 1) % CLOCKWISE_BEARINGS.size]
  end

  def advance
    @pt += BEARINGS[bearing]
  end

  def at(x, y)
    @pt = Pt.new(x, y)
  end

  def coordinates
    pt.to_a
  end
end

class Simulator
  INSTRUCTIONS = {
    "L" => :turn_left,
    "R" => :turn_right,
    "A" => :advance
  }

  def instructions(input)
    input.chars.map { |c| INSTRUCTIONS[c] }
  end

  def place(robot, x:, y:, direction:)
    robot.at(x, y)
    robot.orient(direction)
  end

  def evaluate(robot, input)
    instructions(input).each do |i|
      robot.method(i).call
    end
  end
end
